### ISSUE TITLE
[Google3 Bug]: Illogical keyboard navigation order for FloatingActionButton when placed visually before a long scrollable list

### ISSUE THREAD

**Issue by carriejiang** [flutter/flutter#184077]
### Help us understand the severity of this issue

- [ ] causing severe production issues e.g. malfunctions or data loss
- [ ] blocking next binary release
- [x] blocking a client feature launch within a quarter
- [ ] nice-to-have but does not block a launch within the next quarter

### Steps to reproduce

1.  Create a page with a scaffold and floating action button. The body should contain a long scrollable list of widgets
2.  Using a bluetooth keyboard or screen reader, navigate through the page (using tab on the keyboard or swiping right on the screen reader)

### Expected results

Users should be able to navigate to the "+" (FAB) button first before the list in the body.

### Actual results

Users have to navigate through the long list before the FAB button.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(useMaterial3: false),
      title: 'Flutter Demo',
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title, style: TextStyle(fontFamily: 'ProductSans')),
      ),
      body: Center(
        child: ListView.builder(
          itemCount: 100,
          itemBuilder: (context, index) {
            return ListTile(
              title: Text('Item $index'),
              onTap: () {
                print('Tapped on Item $index');
              },
            );
          },
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => print('Tapped on FAB'),
        tooltip: 'Add',
        child: const Icon(Icons.add),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/user-attachments/assets/8a09aedc-38e5-4246-87c1-1d625b48a90f

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel google3, on macOS 15.7.4 24G517 darwin-arm64 (Rosetta), locale en) [567ms]
    • Framework revision d40e548ef5 (0 days ago), 2026-03-24T00:00:00.000
    • Engine revision d40e548ef5
    • Dart version b08bd0a3ee

[✓] Android toolchain - develop for Android devices (Android SDK version Stable) [558ms]
    • Android SDK at google3
    • Emulator version unknown
    • Platform Stable, build-tools Stable
    • Java binary at: /usr/bin/java
      This JDK was found in the system PATH.
      To manually set the JDK path, use: flutter config --jdk-dir="path/to/jdk".
    • Java version OpenJDK Runtime Environment Zulu25.28+85-CA (build 25+36-LTS)

[✓] Xcode - develop for iOS and macOS (Xcode 26.2) [1,858ms]
    • Xcode at /Applications/[Xcode_26.2.0.17C52.app/Contents/Developer](http://xcode_26.2.0.17c52.app/Contents/Developer)
    • Build 17C52```

</details>

**Comment by Piinks** [comment]
I am working on a fix, but it might take a bit to land as it affects a wide area of the framework (#184730)

In the meantime, I was able to create a workaround to unblock:

```dart
import 'package:flutter/material.dart';
import 'package:flutter/semantics.dart';


void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(useMaterial3: false),
      title: 'Flutter Demo',
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    // Workaround: Wrap the Scaffold in a FocusTraversalGroup with OrderedTraversalPolicy
    // to manually define the navigation sequence.
    return FocusTraversalGroup(
      policy: OrderedTraversalPolicy(),
      child: Scaffold(
        appBar: AppBar(
          title: Text(widget.title, style: const TextStyle(fontFamily: 'ProductSans')),
        ),
        // Assign a HIGHER order to the body content (it will be traversed later)
        body: FocusTraversalOrder(
          order: const NumericFocusOrder(2.0),
          child: Semantics(
            sortKey: const OrdinalSortKey(2.0),
            container: true,
            child: Center(
              child: ListView.builder(
                itemCount: 100,
                itemBuilder: (context, index) {
                  return ListTile(
                    title: Text('Item $index'),
                    onTap: () {
                      print('Tapped on Item $index');
                    },
                  );
                },
              ),
            ),
          ),
        ),
        // Assign a LOWER order to the FloatingActionButton (it will be traversed earlier)
        floatingActionButton: FocusTraversalOrder(
          order: const NumericFocusOrder(1.0),
          child: Semantics(
            sortKey: const OrdinalSortKey(1.0),
            container: true,
            child: FloatingActionButton(
              onPressed: () => print('Tapped on FAB'),
              tooltip: 'Add',
              child: const Icon(Icons.add),
            ),
          ),
        ),
      ),
    );
  }
}

```

**Comment by carriejiang** [comment]
@Piinks Yes, using tab to navigate. I've attached a video of what I'm seeing on my iOS device with both keyboard accessibility and TalkBack (I added a couple more buttons to the app bar just to make sure the app bar items are navigated through first).

https://github.com/user-attachments/assets/e8ef382f-6254-4373-9b5d-6ab4fde2cfbb

This is the code snippet I used:
```
import 'package:flutter/material.dart';
import 'package:flutter/semantics.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(useMaterial3: false),
      title: 'Flutter Demo',
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    // Workaround: Wrap the Scaffold in a FocusTraversalGroup with OrderedTraversalPolicy
    // to manually define the navigation sequence.
    return FocusTraversalGroup(
      policy: OrderedTraversalPolicy(),
      child: Scaffold(
        drawer: Drawer(
          child: ListView(
            padding: EdgeInsets.zero,
            children: const [
              DrawerHeader(
                decoration: BoxDecoration(color: Colors.blue),
                child: Text('Navigation Menu'),
              ),
              ListTile(title: Text('Home')),
              ListTile(title: Text('Settings')),
            ],
          ),
        ),
        appBar: AppBar(
          title: Text(
            widget.title,
            style: const TextStyle(fontFamily: 'ProductSans'),
          ),
          actions: [
            IconButton(
              icon: const Icon(Icons.search),
              onPressed: () {
                print('Search tapped');
              },
            ),
          ],
        ),
        // Assign a HIGHER order to the body content (it will be traversed later)
        body: FocusTraversalOrder(
          order: const NumericFocusOrder(2.0),
          child: Semantics(
            sortKey: const OrdinalSortKey(2.0),
            container: true,
            child: Center(
              child: ListView.builder(
                itemCount: 100,
                itemBuilder: (context, index) {
                  return ListTile(
                    title: Text('Item $index'),
                    onTap: () {
                      print('Tapped on Item $index');
                    },
                  );
                },
              ),
            ),
          ),
        ),
        // Assign a LOWER order to the FloatingActionButton (it will be traversed earlier)
        floatingActionButton: FocusTraversalOrder(
          order: const NumericFocusOrder(1.0),
          child: Semantics(
            sortKey: const OrdinalSortKey(1.0),
            container: true,
            child: FloatingActionButton(
              onPressed: () => print('Tapped on FAB'),
              tooltip: 'Add',
              child: const Icon(Icons.add),
            ),
          ),
        ),
      ),
    );
  }
}
```

**Comment by carriejiang** [comment]
@Piinks Just checking if there were any updates to this?
